### PR TITLE
[SPARK-39607][SQL][DSV2] Distribution and ordering support V2 function in writing

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

Add new feature, make distribution and ordering support V2 function in writing.

Currently, the rule `V2Writes` support converting `ApplyTransform` to `TransformExpression` (unevaluable), this PR makes `V2Writes` supports converting `TransformExpression` to `ApplyFunctionExpression`/`Invoke`/`StaticInvoke` (evaluable).

### Why are the changes needed?

SPARK-33779 introduced API for DSv2 writer to claim distributions and orderings of data before writing, w/ SPARK-34026, Spark can translate `IdentityTransform` to catalyst expression in distributions and orderings expressions.

But for some databases like ClickHouse, which allows table partition defined by an expression, e.g. `PARTITIONED BY num % 10` it's useful to support translating `ApplyTransform` so Spark can organize the data to fit the target storage requirement before writing.

### Does this PR introduce _any_ user-facing change?

Yes, user can use V2 function as partition transform in DSv2 connector.

### How was this patch tested?

UT added.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 1]
```diff
@@ -53,16 +63,27 @@ object DistributionAndOrderingUtils {
...
+          ordering.map(ur => resolveTransformExpression(ur).asInstanceOf[SortOrder]),
+          global = false,
+          queryWithDistribution)
       } else {
         queryWithDistribution
       }
 
-      queryWithDistributionAndOrdering
+      // Apply typeCoercionRules
+      SimpleAnalyzer.execute(queryWithDistributionAndOrdering)
```
Comment (pan3793) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:77]: It's a little bit hacky to re-run the analyzer in optimize phase, but it may be the simplest way

**Comment by pan3793** [comment]
cc @cloud-fan @sunchao @ulysses-you @huaxingao

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 2]
```diff
@@ -17,22 +17,32 @@
...
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction
 import org.apache.spark.sql.connector.distributions._
 import org.apache.spark.sql.connector.write.{RequiresDistributionAndOrdering, Write}
 import org.apache.spark.sql.errors.QueryCompilationErrors
 
 object DistributionAndOrderingUtils {
 
-  def prepareQuery(write: Write, query: LogicalPlan): LogicalPlan = write match {
+  def prepareQuery(
+    write: Write,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: nit: 4 spaces indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala [hunk 3]
```diff
@@ -143,4 +150,54 @@ object V2ExpressionUtils extends SQLConfHelper with Logging {
...
+        }
+    }
+  }
+
+  /**
+   * Check if the input `fn` implements the given `methodName` with parameter types specified
+   * via `argClasses`.
+   */
+  private def findMethod(
+    fn: BoundFunction,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala:192]: nit: 4 spaces indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala [hunk 4]
```diff
@@ -143,4 +150,54 @@ object V2ExpressionUtils extends SQLConfHelper with Logging {
     case V2NullOrdering.NULLS_FIRST => NullsFirst
     case V2NullOrdering.NULLS_LAST => NullsLast
   }
+
+  def resolveV2ScalarFunction(
+    scalarFunc: ScalarFunction[_],
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala:155]: 4 spaces indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala [hunk 5]
```diff
@@ -143,4 +150,54 @@ object V2ExpressionUtils extends SQLConfHelper with Logging {
...
+    arguments: Seq[Expression]
+  ): Expression = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala:157]:
```suggestion
    arguments: Seq[Expression]): Expression = {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 6]
```diff
@@ -53,16 +63,27 @@ object DistributionAndOrderingUtils {
...
-      queryWithDistributionAndOrdering
+      // Apply typeCoercionRules
+      SimpleAnalyzer.execute(queryWithDistributionAndOrdering)
 
     case _ =>
       query
   }
+
+  def resolveTransformExpression(expr: Expression): Expression = expr.transform {
+    case TransformExpression(scalarFunc: ScalarFunction[_], arguments, Some(numBuckets)) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:84]: I can't recall the details now. When do we need to translate v2 trasnform to `TransformExpression`?

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:84]: `TransformExpression` is introduced in SPARK-37377(#35657), to check input partition compatibility in `V2ScanPartitioningAndOrdering`

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:87]: SPARK-37377(#35657) changes the common code used by V2 reading & writing, converting the `Transform` to the `TransformExpression`. 
If wanted, we can skip `TransformExpression` and resolve `Transform` to `ApplyFunctionExpression` / `Invoke` / `StaticInvoke` directly, but this may cause V2 reading & writing to share less code.

**Comment by pan3793** [comment]
The rule `V2Writers` is placed in the optimizing phase but does something maybe should happen in analyzing phase like resolve function and implicit type cast. The current approach is trying to minimize the code change, I'm not sure if it's the right direction.

**Comment by pan3793** [comment]
@cloud-fan @sunchao sorry to bother you again, please take a look when you have time.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 7]
```diff
@@ -17,22 +17,33 @@
...
+import org.apache.spark.sql.connector.catalog.FunctionCatalog
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction
 import org.apache.spark.sql.connector.distributions._
 import org.apache.spark.sql.connector.write.{RequiresDistributionAndOrdering, Write}
 import org.apache.spark.sql.errors.QueryCompilationErrors
 
 object DistributionAndOrderingUtils {
 
-  def prepareQuery(write: Write, query: LogicalPlan): LogicalPlan = write match {
+  def prepareQuery(
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: Hmm I wonder how does the write work with transforms such as bucket. For example, suppose the required distribution is `bucket(col, 100)`, Spark currently will compute the partition (bucket) ID by `murmur_hash(bucket(col, 100)) pmod 100`, so the value of `col` is essentially hashed twice. I'm not sure whether this breaks any assumption from the V2 data source side, or whether it has any effect in the hash key distributions.

**Review by sunchao — COMMENTED** [review]
Looks OK to me. I'm just a bit confused on how the transforms are going to be used in the write path & shuffling data. 

cc @aokolnychyi too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 7 above)
Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: > Spark currently will compute the partition (bucket) ID by `murmur_hash(bucket(col, 100)) pmod 100`

It's only happening on the V1 write path. In V2, the bucket is resolved as `BucketTransform` in analyzing phase, and converted to evaluable catalyst expression `ApplyFunctionExpression`/`Invoke`/`StaticInvoke` here, so I don't think your concern will happen.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 8]
```diff
@@ -17,22 +17,33 @@
...
+      funCatalogOpt: Option[FunctionCatalog]): LogicalPlan = write match {
     case write: RequiresDistributionAndOrdering =>
       val numPartitions = write.requiredNumPartitions()
 
       val distribution = write.requiredDistribution match {
-        case d: OrderedDistribution => toCatalystOrdering(d.ordering(), query)
-        case d: ClusteredDistribution => d.clustering.map(e => toCatalyst(e, query)).toSeq
+        case d: OrderedDistribution =>
+          toCatalystOrdering(d.ordering(), query, funCatalogOpt)
+            .map(ur => resolveTransformExpression(ur).asInstanceOf[SortOrder])
```
[the remark this answers is not in the document]
Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:43]: updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 7 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: Hmm, yes it will be converted to `ApplyFunctionExpression` etc. But in V2 write path, these partition transforms will be used in `RepartitionByExpression`, and then converted into `ShuffleExchangeExec` during physical planning, right? 

Checking the code path, the `ApplyFunctionExpression` / `Invoke` / `StaticInvoke`, etc, will be passed to `RepartitionByExpression` first, as field `partitionExpressions`, and then be passed to `HashPartitioning`, `RangePartitioning` etc, in `HasPartitionExpressions`, and eventually be used in places like `HashPartitioning.partitionIdExpression` where the computation of partition ID I mentioned above happened.

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: Thanks for explaining, educated. After reading the code, I think you are right that "the value of `col` is essentially hashed twice", but I don't think it will bring correctness issues, because it still guarantees that the same bucket value clustered into the same partition. 
One example is Hive bucket. In `V1WritesUtils#getWriterBucketSpec`, both `HiveHash`  and `HashPartitioning#partitionIdExpression` can be used to construct `bucketIdExpression`.

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: I'm just thinking whether we should enforce stronger semantics on the bucketing. Consider we build bucketed table support for DSv2 file source on top on this mechanism, does it mean for a particular row, it could be hashed to a different bucket ID than V1? What if someone wants to first write a bucketed table using V2 and then read back in V1, and perhaps do bucket join with another V1 bucketed table? I wonder if that could cause incorrect results.

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: In V2, the V1 bucketed table is a partitioned table w/ a special `bucket` function (we should give it a special canonical name), `TransformExpression#isSameFunction` will forbid bucket join if they use different bucket functions. So I think there is no correctness issue.

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: But in that case the V1 and V2 bucket tables are incompatible, which is not very desirable right?

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: Given that different storage system usually defines their own hash functions and sharding(bucketing) rules, I don't think we can make bucket tables of all data sources fully compatible w/ each other, e.g. Spark, Hive, Iceberg using the total different hash algorithm for bucketing.

We can make that DS V1 file bucket table compatible w/ V2 file bucket table, and Hive bucket table compatible w/ V2 Hive bucket table, but DS file bucket table can not be compatible w/ Hive(neither V1 nor V2) since they use the different hash algorithm.

In detail, as `V2SessionCatalog` extends `FunctionCatalog`, to make V2 file bucket tables compatible w/ V1, we can introduce and register a `SparkBucket` V2 function using the same hash algorithm V1 in `V2SessionCatalog`. For Hive tables, just register a similar `HiveBucket` V2 function using the Hive hash algorithm in `HiveCatalog`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala [hunk 9]
```diff
@@ -17,22 +17,33 @@
...
-  def prepareQuery(write: Write, query: LogicalPlan): LogicalPlan = write match {
+  def prepareQuery(
+      write: Write,
+      query: LogicalPlan,
+      funCatalogOpt: Option[FunctionCatalog]): LogicalPlan = write match {
     case write: RequiresDistributionAndOrdering =>
       val numPartitions = write.requiredNumPartitions()
 
       val distribution = write.requiredDistribution match {
-        case d: OrderedDistribution => toCatalystOrdering(d.ordering(), query)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:34]: what was the behavior before this PR? do we fail to translate v2 function or we fail at runtime complaining that some expression can't be evaluated?

**Comment by aokolnychyi** [comment]
I'd love to take a look tomorrow as well (not blocking).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 9 above)
Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:34]: The former, the query is going to fail w/ `org.apache.spark.sql.AnalysisException: ${expr} is not currently supported`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 7 above)
Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: I think @sunchao brings a valid point that is easy to overlook. We have to make sure Spark writes to Hive tables in the same way no matter whether the v1 or v2 path is being used.

Would it be correct to say we have this issue because `partitionIdExpression` in `HashPartitioning` is used both for generating bucket IDs in Hive tables as well as for producing partition IDs for writing tasks? Can we use different mechanisms?

Comment (pan3793) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: After SPARK-32709(https://github.com/apache/spark/pull/33432), it allows using different hash expressions for generating partition IDs and bucket IDs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala [hunk 10]
```diff
@@ -143,4 +150,53 @@ object V2ExpressionUtils extends SQLConfHelper with Logging {
     case V2NullOrdering.NULLS_FIRST => NullsFirst
     case V2NullOrdering.NULLS_LAST => NullsLast
   }
+
+  def resolveV2ScalarFunction(
```
[the remark this answers is not in the document]
Comment (pan3793) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala]: yea, naming w/ v2 is redundant, removed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala [hunk 11]
```diff
@@ -48,6 +48,11 @@ case class DataSourceV2Relation(
 
   import DataSourceV2Implicits._
 
+  lazy val funCatalog: Option[FunctionCatalog] = catalog.flatMap {
```
[the remark this answers is not in the document]
Comment (pan3793) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala]: thanks, updated

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 7 above)
Comment (aokolnychyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: > Consider we build bucketed table support for DSv2 file source on top on this mechanism, does it mean for a particular row, it could be hashed to a different bucket ID than V1? What if someone wants to first write a bucketed table using V2 and then read back in V1, and perhaps do bucket join with another V1 bucketed table?

@sunchao, coming back to the use case you mentioned above. I think the bucket ID will be always the same as long as the task writer respects the table bucket spec and Spark shuffles all records that are supposed to land in one bucket to one task. Like @pan3793 said, I suppose V2 file source will request a distribution using a V2 function that would wrap the internal Spark hash function. That should guarantee that all records for the same bucket will land in one task. As long as the task writer uses the correct bucket expression (based on the table definition), we should be good, right?

You are right the partition/task ID during writes for a particular row may be different in V1 and V2 because of double hashing. But does it actually matter, though?

**Comment by cloud-fan** [comment]
In general, this feature looks reasonable, but it's interesting to discuss the behavior of "v2 write required distribution" with this new feature.

Let's assume the required distribution is `ClusteredDistribution`, its doc says
```
/**
 * A distribution where tuples that share the same values for clustering expressions are co-located
 * in the same partition.
 *
 * @since 3.2.0
 */
@Experimental
public interface ClusteredDistribution extends Distribution
```

This means, the clustering expressions are the keys, and Spark makes sure records with the same keys go to the same partition. What Spark does is: for each record, calculate the keys, hash the keys and assign a partition ID for the record based on the hash of the keys.

How can we use this feature to implement bucket writing? We can use the expression (a v2 function) that calculates the bucket ID as the clustering expressions. Then Spark will make sure records with the same bucket ID will be in the same partition. However, the problem of this approach is low parallelism (at most number of buckets).

A different approach is to use the bucket columns as the clustering expressions. Spark will make sure records with the same bucket columns values will be in the same partition. Then the v2 write can require a local sort with bucket id (a v2 function) so that records with the same bucket ID will be grouped together.

That said, I think most users will not use bucket transform as the clustering expressions. If they do, it's there choice and Spark won't do anything wrong.

What do you think? @sunchao @aokolnychyi

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala (hunk 7 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala:33]: Sorry for the late reply here. Yes, I think it should be OK as long as the task writer uses the bucket expression to decide which bucket the input record should go. There will be some extra work to achieve V1 and V2 compatibility but it's not that relevant to this PR now.

**Comment by sunchao** [comment]
> How can we use this feature to implement bucket writing? We can use the expression (a v2 function) that calculates the bucket ID as the clustering expressions. Then Spark will make sure records with the same bucket ID will be in the same partition. However, the problem of this approach is low parallelism (at most number of buckets).

@cloud-fan I think you raised a good point. With the double-hashing mentioned above the parallelism could even be less than the number of buckets due to collision (but I guess this is just a minor thing since the chance is low). Even though the actual number of Spark tasks may be much larger than the number of buckets, most of the tasks will receive empty input in this scenario.

> A different approach is to use the bucket columns as the clustering expressions. Spark will make sure records with the same bucket columns values will be in the same partition. Then the v2 write can require a local sort with bucket id (a v2 function) so that records with the same bucket ID will be grouped together.

This means it now relies on Spark's hash function for bucketing though, which could be different from other engines. I think it would cause compatibility issues, right?

> That said, I think most users will not use bucket transform as the clustering expressions.

Hmm I'm not sure whether this is true. @aokolnychyi may know more from Iceberg side.

**Comment by sunchao** [comment]
In general, let's say a V2 transform function maps a key to a value, given a set of keys, the value space should always be <= the key space. Therefore, it's seems better for Spark to shuffle the records based on the keys, while still uses the V2 transform function provided to determine the actual partition/bucket a record should belong to.

Maybe we can introduce a special case for `KeyGroupedPartitioning` in `ShuffleExchangeExec`, which only consider keys, rather values after V2 transform function evaluation, when computing hash and deciding the partition ID for an input record.

**Comment by cloud-fan** [comment]
> This means it now relies on Spark's hash function for bucketing though, which could be different from other engines.

Let's think about it this way: The v2 data source only needs Spark to local-sort the data by bucket id, which means the required ordering will be a v2 function that generates bucket id. Then the v2 writer generates the bucket id again using the same v2 function during data writing. Or the v2 writer can just use a hash map to keep open file handlers so that Spark doesn't need to sort the data. The extra clustering is only to reduce the number of files we write out.

The Spark hash algorithm only matters when reading bucketed tables and trying to avoid shuffles. I think this is handled well already. Spark will shuffle a v2 table scan again if the other side of the join is a normal table scan with shuffle.

**Comment by sunchao** [comment]
Makes sense. So you mean V2 data source providers can report only the clustered columns through `RequiresDistributionAndOrdering` while still use the V2 transform functions on those columns in the V2 writer tasks.

**Comment by cloud-fan** [comment]
@sunchao yes

**Comment by aokolnychyi** [comment]
> This means, the clustering expressions are the keys, and Spark makes sure records with the same keys go to the same partition. What Spark does is: for each record, calculate the keys, hash the keys and assign a partition ID for the record based on the hash of the keys.
>
> How can we use this feature to implement bucket writing? We can use the expression (a v2 function) that calculates the bucket ID as the clustering expressions. Then Spark will make sure records with the same bucket ID will be in the same partition. However, the problem of this approach is low parallelism (at most number of buckets).
>
> A different approach is to use the bucket columns as the clustering expressions. Spark will make sure records with the same bucket columns values will be in the same partition. Then the v2 write can require a local sort with bucket id (a v2 function) so that records with the same bucket ID will be grouped together.

@cloud-fan, I agree with your summary. It seems to me like a classic trade-off between fewer files (clustering by bucket ID) and better parallelism (local sort by bucket ID). I believe the current Spark API is flexible enough so that data sources can request either of those depending on their internal logic. The third alternative is to leverage an ordered distribution by bucket ID + some other key. In that case, Spark will do skew estimation while determining ranges.

To sum up, I feel the logic in this PR works and the existing API should cover all discussed cases.

What do you think, @cloud-fan @sunchao?

**Comment by aokolnychyi** [comment]
I guess AQE can also help us to coalesce and split too big partitions during writes.

**Comment by pan3793** [comment]
> I guess AQE can also help us to coalesce and split too big partitions during writes.

This should be covered by SPARK-37523.

Thanks @cloud-fan @sunchao @aokolnychyi for reviewing and discussing, and is there anything I can do before this PR gets in?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
